### DISCUSSION TITLE
Restarting Stream Deck app or computer removes button image and text

### DISCUSSION THREAD

**Discussion by Muhomorik** [elgatosf/elgato#106] [Help] [answered]
Hello,

I have a problem that when I set image and text with code, values disappear when I restart the app or computer (ofc).
I can seen the icon and text while computer powers off and until SD app has loaded.
Then it resets to icon in the manifest.

PI

```
    <div class="sdpi-item">
         <div class="sdpi-item-label">setImage</div>
         <button class="sdpi-item-value" id="btn_setImage" onclick="testSetImage()">Sent image</button>
    </div>
```

```
    const imgGreen = 'data:image/png;base64,...'
    const imgRed= 'data:image/png;base64,...'
    
    let imgIndex = 0;
    
    function testSetImage(clickedTab) {
        console.log('testSetGlobals');
    
        let currImage = '';
        if (imgIndex === 0){
            imgIndex = 1;
            currImage = imgGreen;
        } else {
            imgIndex = 0;
            currImage = imgRed;
        }
    
        $PI.sendToPlugin({
            cmd: 'image',
            payload: currImage
        })
    }
```

Plugin (app.js)

```
    $SD.on(`com.iccms.dmitri-1.action-test-1.sendToPlugin`, (jsn) => {
        console.log('Main, sendToPlugin')

        if (jsn.payload.cmd ==='image'){
            $SD.setImage(jsn.context, jsn.payload.payload);
        }
    });
```

Do I need to handle icon and text, set by code somehow at launch? In that case in which event?
I would expect it to be saved and restored.

**Comment by ChekTek** [comment, marked as answer]
Hi @Muhomorik,

That is correct, user set icons and titles are retained on restart, but icons and titles provided by the plugin will need to be set again. There are a few events that can help with this: `willAppear` on the action (this is typically where I handle drawing things) or `systemDidWakeUp` from Stream Deck.

https://docs.elgato.com/sdk/plugins/events-received#willappear
https://docs.elgato.com/sdk/plugins/events-received#systemdidwakeup

The SDK also provides a `connected` event that is sent when the plugin opens the websocket connection to Stream Deck.

All of these could work to help keep the key icon and title up to date.
